File: src/main/python/modular/compute/promotion.py
from typing import Optional

from modular.tree.RootedForest import RootedForest, Node
from modular.compute.MDComputeNode import MDComputeNode, SplitDirection


def promote(tree: RootedForest[MDComputeNode], prob: Node[MDComputeNode]):
    promote_one_direction(tree, prob, SplitDirection.LEFT)
    promote_one_direction(tree, prob, SplitDirection.RIGHT)


def promote_one_direction(tree: RootedForest[MDComputeNode], prob: Node[MDComputeNode], split_type: SplitDirection):
    for c in prob.get_children():
        promote_one_node(tree, c, split_type)


def promote_one_node(tree: RootedForest[MDComputeNode], node: Node[MDComputeNode], split_type: SplitDirection):
    # non-recursive implementation

    if node.first_child is None:
        return

    st = [(False, node), (True, node.first_child)]
    while st:
        is_forward, nd = st.pop()
        if is_forward:
            # forward pass
            if nd.right:
                st += [(True, nd.right)]  # point to the next node
            if nd.data.is_split_marked(split_type):
                assert nd.parent is not None
                if split_type == SplitDirection.LEFT:
                    tree.move_to_before(nd, nd.parent)
                else:
                    tree.move_to_after(nd, nd.parent)
                if nd.first_child is not None:
                    st += [(False, nd), (True, nd.first_child)]  # dig into the children
        else:
            # backward pass
            if nd.is_leaf() and nd.data.is_operation_node():
                tree.remove(nd)
            elif nd.has_only_one_child():
                tree.replace_by_children(nd)
                tree.remove(nd)

File: src/main/cpp/modular/compute/promotion.cpp
#include "MDSolver.hpp"

namespace modular {
namespace compute {

namespace impl {

static void promote_one_node(CompTree &tree, int index, SplitDirection split_type) {
  // non-recursive implementation
  if (tree[index].is_leaf()) return;

  std::stack<std::pair<bool, int>> st;
  st.push({false, index});
  st.push({true, tree[index].first_child});

  while (!st.empty()) {
    auto x = st.top();
    auto nd = x.second;
    st.pop();

    if (x.first) {
      // forward pass
      auto right = tree[nd].right;
      if (tree.is_valid(right)) {
        st.push({true, right});  // point to the next node
      }
      if (tree[nd].data.is_split_marked(split_type)) {
        auto p = tree[nd].parent;
        assert(tree.is_valid(p));

        if (split_type == SplitDirection::LEFT) {
          tree.move_to_before(nd, p);
        } else {
          tree.move_to_after(nd, p);
        }
        if (tree[nd].has_child()) {
          st.push({false, nd});
          st.push({true, tree[nd].first_child});  // dig into the children
        }
      }
    } else {
      // backward pass
      if (tree[nd].is_leaf() && tree[nd].data.is_operation_node()) {
        tree.remove(nd);
      } else if (tree[nd].has_only_one_child()) {
        tree.replace_by_children(nd);
        tree.remove(nd);
      }
    }
  }
}

static void promote_one_direction(CompTree &tree, int index, SplitDirection split_type) {
  for (auto c : tree.get_children(index)) promote_one_node(tree, c, split_type);
}

void promote(CompTree &tree, int prob) {
  TRACE("start: %s", tree.to_string(prob).c_str());

  promote_one_direction(tree, prob, SplitDirection::LEFT);
  promote_one_direction(tree, prob, SplitDirection::RIGHT);

  TRACE("finish: %s", tree.to_string(prob).c_str());
}

}  // namespace impl
}  // namespace compute
}  // namespace modular
